org.eclipse.xtext.xbase/src/org/eclipse/xtext/xbase/validation/XbaseValidator.java

```diff
@@ -1346,7 +1348,29 @@ public void checkPrimitiveComparedToNull(XBinaryOperation binaryOperation) {
 	}
 		
 	protected boolean isLocallyUsed(EObject target, EObject containerToFindUsage) {
-		return readAndWriteTracking.isRead(target) || !XbaseUsageCrossReferencer.find(target, containerToFindUsage).isEmpty();
+		if (readAndWriteTracking.isRead(target)) {
+			return true;
+		}
+		Collection<Setting> usages = XbaseUsageCrossReferencer.find(target, containerToFindUsage);
+		// for non-private members it is enough to check that there are usages
+		if (!(target instanceof JvmOperation) || ((JvmOperation)target).getVisibility()!=JvmVisibility.PRIVATE) {
+			return !usages.isEmpty();
+		} else {
+			// for private members it has to be checked if all usages are within the operation
+			EObject targetSourceElem = associations.getPrimarySourceElement(target);
+			long featureCallsWithinTarget = usages.stream()
+				// filter all feature calls to target
+				.map(s -> s.getEObject())
+				.filter(it -> it instanceof XFeatureCall)
+				.map(XFeatureCall.class::cast)
+				.filter(fc->fc.getFeature()==target)
+				// filter all feature calls which are contained in the same operation
+				.filter(fc->Iterables.contains(EcoreUtil2.getAllContainers(fc), targetSourceElem))
+				.count()
+				;
+			// if not all usages are recursive calls return true
+			return usages.size() != featureCallsWithinTarget;
```

Comment:
The Java8 style stream code above always traverses all elements. A simple for loop would be more efficient, since then you can early exit with `return true` as soon as you find an element that is not contained in the operation.